File: problems/3143_maximum_points_inside_the_square/solution.py
from typing import List


class Solution:
    def maxPointsInsideSquare(self, points: List[List[int]], s: str) -> int:
        from collections import defaultdict
        tag_dists = defaultdict(list)
        for i, (x, y) in enumerate(points):
            d = max(abs(x), abs(y))
            tag_dists[s[i]].append(d)
        limit = float('inf')
        for tag, dists in tag_dists.items():
            dists.sort()
            if len(dists) >= 2:
                limit = min(limit, dists[1])
        res = 0
        for tag, dists in tag_dists.items():
            if dists[0] < limit:
                res += 1
        return res

File: problems/3143_maximum_points_inside_the_square/Solution.java
import java.util.*;

class Solution {
    public int maxPointsInsideSquare(int[][] points, String s) {
        Map<Character, List<Integer>> map = new HashMap<>();
        for (int i = 0; i < points.length; i++) {
            int d = Math.max(Math.abs(points[i][0]), Math.abs(points[i][1]));
            map.computeIfAbsent(s.charAt(i), k -> new ArrayList<>()).add(d);
        }
        int limit = Integer.MAX_VALUE;
        for (var e : map.entrySet()) {
            Collections.sort(e.getValue());
            if (e.getValue().size() >= 2) limit = Math.min(limit, e.getValue().get(1));
        }
        int res = 0;
        for (var e : map.entrySet())
            if (e.getValue().get(0) < limit) res++;
        return res;
    }
}
